tests/frontend/io/test_io_xir.py

```diff
@@ -0,0 +1,646 @@
+
+        xir_prog = io.to_xir(sf_prog)
+        assert [(stmt.name, stmt.params, stmt.wires) for stmt in xir_prog.statements] == expected
+
+    def test_free_par_str(self):
+        """Test a FreeParameter with some transformations converts properly"""
+        sf_prog = Program(2)
+        r, alpha = sf_prog.params("r", "alpha")
+        with sf_prog.context as q:
+            ops.Sgate(r) | q[0]
+            ops.Zgate(3 * pf.log(-alpha)) | q[1]
+
+        xir_prog = io.to_xir(sf_prog)
+
+        expected = [("Sgate", ["r", 0.0], (0,)), ("Zgate", ["3*log(-alpha)"], (1,))]
+        assert [(stmt.name, stmt.params, stmt.wires) for stmt in xir_prog.statements] == expected
+
+    def test_tdm_program(self):
+        sf_prog = TDMProgram(2)
+
+        with sf_prog.context([1, 2], [3, 4], [5, 6]) as (p, q):
+            ops.Sgate(0.7, 0) | q[1]
+            ops.BSgate(p[0]) | (q[0], q[1])
+            ops.Rgate(p[1]) | q[1]
+            ops.MeasureHomodyne(p[2]) | q[0]
+
+        xir_prog = io.to_xir(sf_prog)
+
+        expected = [
+            ("Sgate", [0.7, 0], (1,)),
+            ("BSgate", ["p0", 0.0], (0, 1)),
+            ("Rgate", ["p1"], (1,)),
+            ("MeasureHomodyne", {"phi": "p2"}, (0,)),
+        ]
+        assert [(stmt.name, stmt.params, stmt.wires) for stmt in xir_prog.statements] == expected
+
+
+class TestXIRtoSFConversion:
+    """Tests for the io.to_program utility function for XIR programs"""
+
+    def test_empty_program(self):
+        """Test empty program raises error"""
+        xir_prog = xir.Program()
+
+        with pytest.raises(
+            ValueError, match="is empty and cannot be transformed into a Strawberry Fields program"
+        ):
+            io.to_program(xir_prog)
+
+    def test_gate_not_defined(self):
+        """Test unknown gate raises error"""
+        xir_prog = xir.Program()
+        xir_prog.add_statement(xir.Statement("np", [1, 2, 3], (0,)))
+
+        with pytest.raises(NameError, match="operation 'np' not defined"):
+            io.to_program(xir_prog)
+
+    def test_gate_no_arg(self):
+        """Test gate with no argument converts"""
+        xir_prog = xir.Program()
+        xir_prog.add_statement(xir.Statement("Vacuum", [], (0,)))
+
+        sf_prog = io.to_program(xir_prog)
+
+        assert len(sf_prog) == 1
+        assert sf_prog.circuit
+        assert sf_prog.circuit[0].op.__class__.__name__ == "Vacuum"
+        assert sf_prog.circuit[0].reg[0].ind == 0
+
+    def test_gate_arg(self):
+        """Test gate with arguments converts"""
+        xir_prog = xir.Program()
+        xir_prog.add_statement(xir.Statement("Sgate", [0.54, 0.12], (0,)))
+
+        sf_prog = io.to_program(xir_prog)
+
+        assert len(sf_prog) == 1
+        assert sf_prog.circuit
+        assert sf_prog.circuit[0].op.__class__.__name__ == "Sgate"
+        assert sf_prog.circuit[0].op.p[0] == 0.54
+        assert sf_prog.circuit[0].op.p[1] == 0.12
+        assert sf_prog.circuit[0].reg[0].ind == 0
```

Comment:
What do you think about serializing the SF program to a Blackbird program and comparing that instead?

Reply:
I'd probably prefer having it like this. It's more unit-testy, and doesn't rely on the Blackbird serialization method.

Reply:
That's true although I would argue that this approach is also creating a larger barrier for adding new tests (since more effort and typing is required). Just something to keep in mind!

Reply:
I actually did it slightly differently for the [new tests that I added here](https://github.com/XanaduAI/strawberryfields/blob/1f252b93b21644c6c5917235fd7e52f0b3603bc8/tests/frontend/io/test_io_xir.py#L485-L515). These ones were just copied from the Blackbird tests, so you could say the barrier had already been added, but I agree that they're not the simplest ones.